### PR TITLE
:label: add new endpoint specs

### PR DESCRIPTION
**Pull request by bcesur**

### CODE REVIEW

File: api/v1/api.proto [hunk 1]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  // Start time of the query
+  string start_date = 1;
+  // End time of the query
+  string end_date = 2;
+  // Cloud provider that you want to query
+  string provider = 3;
+  // Accounts that are used
+  string account = 4;
+  // Service of the cloud provider
+  string region = 5;
```
Comment (leonardpahlke) [inline api/v1/api.proto:58]: reuse `Location`
```suggestion
Location location = 5;
```

File: api/v1/api.proto [hunk 2]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  // Request status information
+  carbonaut.api.v1.util.Status status = 2;
+}
+
+// List Footprint
+message ListFootprintForProjectRequest {
+  // Start time of the query
+  string start_date = 1;
+  // End time of the query
+  string end_date = 2;
```
Comment (leonardpahlke) [inline api/v1/api.proto:77]: message `UsageTime`
```suggestion
UsageTime recorded_time_span = 1;
```

File: api/v1/api.proto [hunk 3]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  repeated string services = 1;
+  // Request status information
+  carbonaut.api.v1.util.Status status = 3;
+}
+// List Emissions
+message GetEmissionForProjectRequest {
+  // Start time of the query
+  string start_date = 1;
+  // End time of the query
+  string end_date = 2;
```
Comment (leonardpahlke) [inline api/v1/api.proto:52]: message `UsageTime`
```suggestion
UsageTime recorded_time_span = 1;
```

File: api/v1/api.proto [hunk 4]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  // Cloud provider that you want to query
+  string provider = 3;
+  // Accounts that are used
+  string account = 4;
```
Comment (leonardpahlke) [inline api/v1/api.proto:56]: new message `Provider`? (needs to get created)
```suggestion
Provider provider = 3;
```

File: api/v1/model/model.proto [hunk 5]
```diff
@@ -45,7 +45,36 @@ message Emission {
...
+    string cloud_provider = 1;
+    // Account of the gathered information
+    string account = 2;
+    // Readable name for the used service
+    string service_description = 3;
+    // Recorded Emissions
+    string emission = 4;
+    // Date
+    string timestamp = 5;
+}
```
Comment (leonardpahlke) [inline api/v1/model/model.proto:60]: ? we already have a message Emission & CarbonRecord why is this needed?
```suggestion
```

File: api/v1/api.proto [hunk 6]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  // Service of the cloud provider
+  string region = 5;
```
Comment (leonardpahlke) [inline api/v1/api.proto:83]:
```suggestion
Location location = 5;
```

File: api/v1/api.proto [hunk 7]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+// List Footprint
+message ListFootprintForProjectRequest {
+  // Start time of the query
+  string start_date = 1;
+  // End time of the query
+  string end_date = 2;
+  // Cloud provider that you want to query
+  string provider = 3;
+  // Accounts that are used
+  string account = 4;
```
Comment (leonardpahlke) [inline api/v1/api.proto:81]: I think we can define a message new message here to abstract `provider` with a list of account ids (maybe more fields?) 
```suggestion
Provider provider = 3;
```
(`Provider` needs to be defined)

File: api/v1/api.proto [hunk 8]
```diff
@@ -17,9 +17,88 @@ service EmissionData {
...
+  string groupBy = 6;
+  // The maximum number of items to return.
+  int32 page_size = 7;
+  // The next_page_token value returned from a previous List request, if any.
+  string page_token = 8;
+}
+
+message ListFootprintForProjectResponse {
+  // Carbon footprint data of the project by given dates
+  repeated carbonaut.api.v1.model.Footprint footprints = 1;
```
Comment (leonardpahlke) [inline api/v1/api.proto:94]: maybe reuse `Emissions` ?
```suggestion
repeated carbonaut.api.v1.model.Emission emissions = 1;
```

File: api/v1/model/model.proto [hunk 9]
```diff
@@ -45,7 +45,36 @@ message Emission {
...
+    string cloud_provider = 2;
+    // Account of the gathered information
+    string account = 3;
+    // Recorded Emissions
+    Emission emission = 4;
+    // Readable name for the used service
+    string service_description = 5;
+    // GEO Location
+    Location location = 6;
+}
```
Comment (leonardpahlke) [inline api/v1/model/model.proto:76]: I think we can reuse the `Emission` message at this point.
```suggestion
message ProjectEmission {
  Provider provider = 1;
  repeated Emission emissions = 2;
  // more?
}
```

**Review by leonardpahlke — CHANGES_REQUESTED** [review]
I think we can connect both versions a bit more and find ways to make the model more generic and easy to use. Probably we need to rearrange some things in the first version (which I created a few days ago) as well to make that work. I think we can keep an eye on emissions now (maybe rephrase some things that are called carbon to emissions) and later we can also add endpoints for energy data and so on :) Thanks @bcesur

File: api/v1/model/model.proto (hunk 5 above)
Comment (bcesur) [reply api/v1/model/model.proto:60]: Yeah when I was creating the API spec I thought we would need something like this but converting the stuff to .proto I realized that we have similar stuff already. Added this to have your opinion. Removing it.

File: api/v1/model/model.proto [hunk 10]
```diff
@@ -45,7 +45,11 @@ message Emission {
     // Carbon footprint caused by location
     CarbonRecord carbon_footprint_estimation_location_average = 6;
 }
-  
+
+message Provider {
+    string cloud_provider = 1;
```
Comment (leonardpahlke) [inline api/v1/model/model.proto:50]: i think we can define an `enum` with a selection `AWS, GCP, KUBERNETES, SELF_HOSTED, ... other clouds ...`

File: api/v1/api.proto [hunk 11]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
...
+
+// List Services
+message ListServicesForProjectRequest {}
+
+message ListServicesForProjectResponse {
+  // List of services that are used
+  repeated string services = 1;
+  // Request status information
+  carbonaut.api.v1.util.Status status = 3;
+}
```
Comment (leonardpahlke) [inline api/v1/api.proto:46]: what is a service? ITResource? or smth different?

File: api/v1/api.proto [hunk 12]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
...
+
+message ListAccountsForProjectResponse {
+  // List of accounts that are used in the provider
+  repeated string accounts = 1;
+  // Request status information
+  carbonaut.api.v1.util.Status status = 3;
+}
+
+// List Services
+message ListServicesForProjectRequest {}
```
Comment (leonardpahlke) [inline api/v1/api.proto:39]: empty?

File: api/v1/api.proto [hunk 13]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
...
+  rpc ListFootprint (ListFootprintForProjectRequest) returns (ListFootprintForProjectResponse) {}
+  rpc GetEmission (GetEmissionForProjectRequest) returns (GetEmissionsForProjectResponse) {}
+  rpc ListServices (ListServicesForProjectRequest) returns (ListServicesForProjectResponse) {}
+  rpc ListAccounts (ListAccountsForProjectRequest) returns (ListAccountsForProjectResponse) {}
+
   rpc ListITResourcesForProject (ListITResourcesForProjectRequest) returns (ListITResourcesForProjectResponse) {}
 }
 
+// List Accounts
+message ListAccountsForProjectRequest {}
```
Comment (leonardpahlke) [inline api/v1/api.proto:29]: empty?

File: api/v1/api.proto [hunk 14]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
...
+
+message ListAccountsForProjectResponse {
+  // List of accounts that are used in the provider
+  repeated string accounts = 1;
```
Comment (leonardpahlke) [inline api/v1/api.proto:33]: maybe return `Provider`?

File: api/v1/api.proto [hunk 15]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
 
   // rpc ListEnergyDataForITResource (ListEnergyDataForITResourceRequest) returns (ListEnergyDataForITResourceResponse) {}
 
+  rpc ListFootprint (ListFootprintForProjectRequest) returns (ListFootprintForProjectResponse) {}
+  rpc GetEmission (GetEmissionForProjectRequest) returns (GetEmissionsForProjectResponse) {}
+  rpc ListServices (ListServicesForProjectRequest) returns (ListServicesForProjectResponse) {}
+  rpc ListAccounts (ListAccountsForProjectRequest) returns (ListAccountsForProjectResponse) {}
```
Comment (leonardpahlke) [inline api/v1/api.proto:23]:
```suggestion
rpc ListProviderAccounts (ListProviderAccountsRequest) returns (ListProviderAccountsResponse) {}
```

File: api/v1/api.proto [hunk 16]
```diff
@@ -17,9 +17,79 @@ service EmissionData {
...
+
```
Comment (leonardpahlke) [inline api/v1/api.proto:24]:
```suggestion
rpc ListProviders (ListProvidersRequest) returns (ListProvidersResponse) {}

```
new endpoint to list configured providers?

**Review by leonardpahlke — CHANGES_REQUESTED** [review]

File: api/v1/api.proto (hunk 11 above)
Comment (bcesur) [reply api/v1/api.proto:46]: So the name of the service that generated the emission. For AWS it can be EC2 or for GCP it can be cloud engine. It is actually an ITResource.

File: api/v1/api.proto (hunk 12 above)
Comment (bcesur) [reply api/v1/api.proto:39]: I was thinking in the REST way here. To parse the user's auth token then get the IAM information. This is why I left it empty. It's the same thing for below

**Review by leonardpahlke — APPROVED** [review]
lgtm from my side

cc @christopherfriedrich @mkorbi
